File: hack/windows/common/get_pod_dm.py
#!/bin/python

import sys
import leveldb
import json

db = leveldb.LevelDB('/var/lib/hyper/lib/hyper.db')

# list all key and value
#for key, value in db.RangeIter():
#        print (key, value);

# check argument
if len(sys.argv) != 3:
  print "%s <pod-id> <property>" % sys.argv[0]
  print "<property>: hostname | device | vmid"
  sys.exit(1)  

# genereate key
pod_id = sys.argv[1]
key = "SB-%s" % pod_id
#print "Get value of key: %s" % (key)

# get property
property_name = sys.argv[2]
if property_name not in ["hostname", "device", "vmid"]:
  print "unknown property %s" % property_name
  sys.exit(2)

try:
  # get value of key
  value = db.Get(key)
  #print  str(value[value.find("{"):])
except:
  print "can not get value of key:%s" % key
  sys.exit(3)

try:
  d = json.loads(str(value[value.find("{"):]))
  if property_name == "hostname":
    if d["VmSpec"] and d["VmSpec"]["hostname"]:
      print "%s" % d["VmSpec"]["hostname"]
  elif property_name == "device":
    for i in d["VolumeList"]:
      if i["Format"] == "raw":
        print "%s" % i["Name"]
  elif property_name == "vmid":
    print "%s" % d["Id"]
  else:
    print "unknow property: %s" % property_name
    sys.exit(4)
except:
  print "get property from json failed:%s" % str(value[value.find("{"):])
  sys.exit(5)

File: hack/windows/common/get_pod_dm.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/syndtr/goleveldb/leveldb"
	"github.com/syndtr/goleveldb/leveldb/opt"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Printf("%v <pod-id>", os.Args[0])
		os.Exit(1)
	}

	db, err := leveldb.OpenFile("/var/lib/hyper/lib/hyper.db.bak", &opt.Options{ReadOnly: true})
	if err != nil {
		panic(fmt.Sprintf("Open hyper.db failed!:%v", err.Error()))
	}
	defer db.Close()

	/*
		iter := db.NewIterator(nil, nil)
		for iter.Next() {
			// Remember that the contents of the returned slice should not be modified, and
			// only valid until the next call to Next.
			key := iter.Key()
			value := iter.Value()
			fmt.Printf("\n\nkey:%q\nvalue:%q",key, value)
		}
		iter.Release()
		err = iter.Error()
		if (err != nil) {
			panic(fmt.Sprintf("Close hyper.db failed!:%v", err.Error()));
		}
	*/

	key := fmt.Sprintf("SB-%v", os.Args[1])
	//fmt.Printf("key: %v\n", key)
	value, err := db.Get([]byte(key), nil)
	if err != nil {
		panic(fmt.Sprintf("Key %v not found(Error:%v)\n", key, err))
	}
	tmp := string(value[:])
	jsonStr := tmp[strings.Index(tmp, "{"):]
	//fmt.Printf("data(%T):%q", jsonStr, jsonStr)

	//convert jsonStr to json
	var data map[string]interface{}
	if err := json.Unmarshal([]byte(jsonStr), &data); err != nil {
		panic(fmt.Sprintf("Can not convert string(%v) to json(Error:%v)\n", jsonStr, err))
	}
	//fmt.Printf("(%T) %v\n", data["VolumeList"], data["VolumeList"])

	item := data["VolumeList"]
	if rec, ok := item.([]interface{}); ok {
		for _, val := range rec {
			if val.(map[string]interface{})["Format"] == "raw" {
				fmt.Printf("%v", val.(map[string]interface{})["Name"])
			}
		}
	} else {
		fmt.Printf("item not a map[string]interface{}: %v\n\n", item)
	}
}
